### PR TITLE
HPCC-30848 Ensure globals are available in bare-metal esp

### PR DESCRIPTION
**Pull request by jakesmith**
## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [ ] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

File: esp/platform/espcfg.cpp
Comment (wangkx) [inline esp/platform/espcfg.cpp:414]: Just curious. Did old code work before? Wondering why not using m_envpt->queryPropTree("global/cost") .

File: esp/platform/espp.cpp
Comment (wangkx) [inline esp/platform/espp.cpp:457]: Curious again. Why not pass the cfgfile to the loadConfiguration like other components (for example, dfuserver: globals.setown(loadConfiguration(defaultYaml, argv, "dfuserver", "DFUSERVER", "dfuserver.xml", nullptr)))?

File: esp/platform/espp.cpp
Comment (wangkx) [inline esp/platform/espp.cpp:465]: The replaceComponentConfig made me confused. I thought it only replace the ComponentConfig(). I guess that the replaceComponentConfig means replacing the config for an HPCC component (including its GlobalConfig and its ComponentConfig.

**Review by wangkx — APPROVED** [review]
@jakesmith the changes look good although I do leave questions/comments. I tested the getRemoteStorage().

**Comment by ghalliday** [comment]
@jakesmith I merged since it was approved, but I wonder if kevin's comment isn't better.  

> Curious again. Why not pass the cfgfile to the loadConfiguration like other components (for example, dfuserver: globals.setown(loadConfiguration(defaultYaml, argv, "dfuserver", "DFUSERVER", "dfuserver.xml", nullptr)))?

I should have let you answer before merging.  As it stands the global properties are not picked up by anything (e.g., tracing) which does not monitor for updates to the configuration.

File: esp/platform/espp.cpp
Comment (jakesmith) [reply esp/platform/espp.cpp:457]: it could, but it wouldn't be in the standard format, because the config that would be instantiated in component config is very different from the getComponentConfig format ubiquitous elsewhere
i.e. esp.xml is effectively a selective copy of the environment tree.
And, nothing in the esp code base relies on getComponentConfig* in esp, and can't do so as it stands, because before and after this change it is empty (well populated with defaultYaml only).
It could instantiate the config to the esp.xml config, but generic helper code that expects getComponentConfig to follow certain norms, wouldn't work.

I don't know there are any places in bare-metal in esp that rely on that though..

It is related to when cloud config came along, TonyF, refactored the config loading code in esp, so that at runtime, it builds a <Environment> based IPT from the esp specific 'esp:' section (that's what buildApplicationLegacyConfig is doing).
I think this was to avoid having to change much of the esp code to use the new config scheme.

Ideally, it would have been the other way around, so that legacy esp was translated into a standarized form that adhered to values.schema.json, and all esp code used it and getComponentConfig consistently - that way generic code elsewhere would also make sense in esp.
Definitely beyond the scope of this PR though.

For the most part, I think as long as bare-metal config options are consistently made available via getGlobalConfig, it's not a significant issue.

File: esp/platform/espp.cpp
Comment (jakesmith) [reply esp/platform/espp.cpp:465]: yes, the method can't replace one without replacing the other, so here Ionly want to replace the global, so am passing  the existing component config and the 'new' global (from env).

**Review by jakesmith — COMMENTED** [review]
@wangkx - see replies.

**Comment by jakesmith** [comment]
> As it stands the global properties are not picked up by anything (e.g., tracing) which does not monitor for updates to the configuration.

global/tracing (if defined under Software/Globals) should be picked up after this change, i.e. anything calling getGlobalConfig*() asking for "tracing" should see it.

You're right though, as the esp config is structured config changes to esp (both bare-metal and cloud) are not going to be propagated. Beyond the scope of this PR, but should be revisited. I've opened a new JIRA: https://track.hpccsystems.com/browse/HPCC-31029

**Comment by wangkx** [comment]
Thanks for the replies.

**Pull request merged**
